### ISSUE TITLE
How to update runtime_block_states.dat file with runtime_block_states.dat.dump.txt?

### ISSUE THREAD

**Issue by good777LUCKY** [GameModsBR/PowerNukkit#329]
# 🤷‍♀️ I have a question

How to update `runtime_block_states.dat` file with `runtime_block_states.dat.dump.txt`?

**Comment by joserobjr** [comment]
It's complicated and risky, doing it wrong could cause block, chunk or level damages. It's also not easily mergeable, so I'm assigned to update those files.

But I'm going to sumarize to spread knowledgement.

## [runtime_block_states.dat](https://github.com/GameModsBR/PowerNukkit/blob/bleeding/src/main/resources/runtime_block_states.dat)
This file is managed by NukkitX devs, please, don't change it otherwise we would be hard to know what was changed and to merge, I used to edit it but I reversed the changes and moved them to the `runtime_block_states_overrides.dat`. This file is basically extracted from the [Bedrock Edition Server Software](https://www.minecraft.net/en-us/download/server/bedrock/) by packet inspection using softwares like [ProxyPass](https://github.com/GameModsBR/Bedrock-ProxyPass) and [Wireshark](https://www.wireshark.org/), most data is present in the `StartGamePacket` but requires a binary understanding of how the [RakNet network](https://github.com/GameModsBR/Nukkit-Network) works and [the structure of the packets](https://wiki.vg/Bedrock_Protocol) (note: this link is outdated most of the time). Minecraft used to use numeric IDs to represent block states (named meta, or block damage in Nukkit), it's no longer the case, they moved that system to a key-pair based block property system, similar to the current Java Edition. When NukkitX updates this file, they also add mappings to the old numeric id representation of the state as a `LegacyStates` list in the NBT structure to workaround this limitation. NukkitX's legacy states are also limited to 4 bits, if a new data can't be stored in 4 bits, then, they will drop detailed or states with no visual differences. PowerNukkit increased this limit to 6 bits (but will downgrade back to 4 at 1.4.0.0-PN to recover compatibility with some plugins) but also have a technic to store up to 32 bits of data as of 1.3.0.0-PN (the old 6 bits-data will be converted to this new system).

## [runtime_block_states_overrides.dat](https://github.com/GameModsBR/PowerNukkit/blob/bleeding/src/main/resources/runtime_block_states_overrides.dat)
As explained above, what is inside this file takes priority over what is inside `runtime_block_states.dat`. I'm assigned to this file and I need to approve or apply all changes to it to avoid conflicts or level inconsistencies. I update this file by testing the game behaviour, searching for patterns, standards and data from various sources like the game's decompiled code, other server implementations, other similar blocks, etc. It's not an easy task because I can't just create my own values, they must match the game as close as possible. Since the game has moved away from the numeric data, it's getting a lot harder to find the right values. Once I have the values, I patch the file by creating a single-use java file to do the binary changes. I'm currently doing it using the [OverridesUpdater](https://github.com/GameModsBR/PowerNukkit/blob/bleeding/src/test/java/test/OverridesUpdater.java) class and I'm committing the changes to keep track on how I have applied such change.

## [runtime_block_states.dat.dump.txt](https://raw.githubusercontent.com/GameModsBR/PowerNukkit/bleeding/runtime_block_states.dat.dump.txt)
It exists for the following reasons: 1) It's hard and annoying to inspect the `runtime_block_states.dat` and the `runtime_block_states_overrides.dat` file even using NBT viewers, people need to see the state <-> meta link easily to implement the blocks correctly. 2) I like to track the changes and Git can't show what has changed in a binary file properly.
Changing it causes no effects to the server in anyway, it's only a documentation file. To generate/update it, just run the [RuntimeBlockStateDumper](https://github.com/GameModsBR/PowerNukkit/blob/bleeding/src/test/java/cn/nukkit/RuntimeBlockStateDumper.java) class.

Here are some content examples:
```
minecraft:cobblestone_wall;wall_connection_type_east=none;wall_post_bit=1;wall_connection_type_south=none;wall_connection_type_west=short;wall_connection_type_north=tall;wall_block_type=mossy_cobblestone
[]
[139:4673]
```
The first line fully describes the state which the current bedrock edition version understand
The first brackets represents the numeric ids which was mapped originally by NukkitX's `runtime_block_states.dat`, if it's empty then it will never be displayed there.
The second brackets represents the numeric ids which are mapped by the `runtime_block_states_overrides.dat`, both brackets uses the `block-id:block-meta` syntax. `139` are cobblestone walls and `4673` represent the state (it's using the hyper-id system)

```
minecraft:wood;wood_type=acacia;stripped_bit=0;pillar_axis=y
[162:12, 162:14, 162:15, 467:4]
```
The second line maps 4 numeric states to the same actual bedrock state. Notice that the last number maps a different block id than the first three, that happens because Mojang moved that state to an entire new block (was a subset of 162, which is WOOD2, and now is on it's own block 467, which is WOOD_BARK), that's how we express this situations. Both block implementation must be aware of these states and do migrations gradually (for example, dropping the new block instead of the old one).
Also notice that this segment has only one bracket group, that means that there are no overrides mapping this state, so the original from NukkitX is being used, which is sufficient. If there were a second empty bracket, the overrides would be incorrectly removing the runtime ids and these states would be visible as an "UPDATE!" block.

```
minecraft:warped_button;button_pressed_bit=1;facing_direction=3
[]
```
I picked this example to show how a state looks like when it has not been configured yet by nobody, it shows the Bedrock Edition state but has no legacy id mappings. This state will never be displayed (unless somebody, may be me or NukkitX, maps a numeric state). If you need a state which is not mapped, please, let me know and I'll prioritize.

## [block-states.ini](https://github.com/GameModsBR/PowerNukkit/blob/bleeding/block-states.ini) and [simple-blocks-nukkit.txt](https://github.com/GameModsBR/PowerNukkit/blob/bleeding/simple-blocks-nukkit.txt)
Basically the same as `runtime_block_states.dat.dump.txt`, they are just documentation files and causes no impact. I did that because I'm also working on some world conversion softwares and I need that kind of information and because the dump file is too verbose, sometimes we just need to see the data condensed with less details.
To update both files you just need to run the [SimpleBlocksReader](https://github.com/GameModsBR/PowerNukkit/blob/bleeding/src/test/java/test/SimpleBlocksReader.java) class.

**Comment by good777LUCKY** [comment]
Thank you!
